content/blogposts/2024/anatomy-of-a-coding-assistant.md

```diff
@@ -0,0 +1,157 @@
+
+---
+
+## Context sources outside of your codebase
+
+In addition to automatic context fetching, we've found that providing a way for the user to manually include context is also helpful. We've done this by letting the user ask questions about specific files by @-mentioning them. More recently, we've realized that there is more context the user sometimes wants to include outside of the codebase, such as documentation, web pages, Slack threads, Linear issues, and Notion documents. We've created an open source protocol called [OpenCtx](https://openctx.org/) to make it super easy to bring in context from any source.
+
+[![20_tweet](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/20_tweet.png)](https://x.com/beyang/status/1792616243742244974)
+
+
+---
+
+## Context fetching for autocomplete
+
+Autocomplete provides code suggestions as you code in your editor. For this feature, we want it to feel as instantaneous as possible because we don’t want to break the flow of the programmer. At the same time, we want to quickly find and identify which code snippets to include as part of the context to provide high-quality suggestions.
+
+![13_autocomplete](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/13_autocomplete.png)
+
+For this, we look at a few sources of information: the cursor position, the surrounding code, and the code graph (a code graph is a representation of the relationships and structures within a codebase, mapping entities such as classes and methods to show how they are interconnected). We use the current cursor position within the code graph to determine if the user wants a single-line suggestion or a multi-line suggestion. Once we determine that, we add more context by looking through recent files and open tabs. Within those, we find code snippets related to the code you're currently writing.
+
+![14_autocomplete2](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/14_autocomplete2.png)
+
+For autocomplete, we don't currently use embeddings. Instead, we consider all relevant chunks from the current file, surrounding file, and recently edited files. We compare each of these chunks to the current code being written using something called the Jaccard similarity coefficient. This provides a very fast retrieval mechanism over this set of files, which is crucial for fitting into the latency budget that autocomplete requires. If you’re curious about this, you can check [this document](https://www.notion.so/How-we-find-similar-code-snippets-for-autocomplete-9eaf5ab9179a4e799952a7f12c94dd6d?pvs=21) for more information about it, and you can even dig into the source code because much of Cody is open source.
```

Comment:
```suggestion
For autocomplete, we don't currently use embeddings. Instead, we consider all relevant chunks from the current file, surrounding file, and recently edited files. We compare each of these chunks to the current code being written using something called the [Jaccard similarity coefficient](https://en.wikipedia.org/wiki/Jaccard_index). This provides a very fast retrieval mechanism over this set of files, which is crucial for fitting into the latency budget that autocomplete requires. If you’re curious about this, you can check [this document](https://www.notion.so/How-we-find-similar-code-snippets-for-autocomplete-9eaf5ab9179a4e799952a7f12c94dd6d?pvs=21) for more information about it, and you can even dig into the source code because much of Cody is open source.
```